refactor: wait `this.load` using futures instead of callbacks

sapphi-red opened the pull request:
The code for `this.load` waited for the `markModuleLoaded` callback to be called. But this should be able to simplify by using a future.
This PR refactors the code for `this.load` to wait for a future instead. This should also improve the perf a bit because we no longer have to call `markModuleLoaded` for each module.
I wonder if I'm missing something.

refs #4156, #2777, #2772, #2687

sapphi-red commented:
* **#6222** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6222?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6222?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6222).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hyf0 commented:
### Merge activity

* **Sep 16, 6:10 AM UTC**: The merge label 'graphite: merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown) once it meets the requirements.
* **Sep 16, 6:10 AM UTC**: `hyf0` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).
* **Sep 16, 6:19 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.dev/merges?org=rolldown&repo=rolldown).

sapphi-red committed:
refactor: wait `this.load` using futures instead of callbacks (#6222)

The code for `this.load` waited for the `markModuleLoaded` callback to be called. But this should be able to simplify by using a future.
This PR refactors the code for `this.load` to wait for a future instead. This should also improve the perf a bit because we no longer have to call `markModuleLoaded` for each module.
I wonder if I'm missing something.

refs #4156, #2777, #2772, #2687
